### PR TITLE
Instrument proctelemetry.ProcessMetrics

### PR DESCRIPTION
**Pull request by moh-osman3**
Description:

This PR instruments proctelemetry.ProcessMetrics with otel go. This is a followup PR that is based on similar changes made for obsreport.Receiver: https://github.com/open-telemetry/opentelemetry-collector/pull/6222

Link to tracking Issue: Part of https://github.com/open-telemetry/opentelemetry-collector/issues/816

Testing:
Built otelcorecol binary and checked metrics endpoint. Tested with and without featuregate.
```
% curl localhost:8888/metrics
# HELP otelcol_process_cpu_seconds_total Total CPU user and system time in seconds
# TYPE otelcol_process_cpu_seconds_total counter
otelcol_process_cpu_seconds_total 27750
# HELP otelcol_process_memory_rss_total Total physical memory (resident set size)
# TYPE otelcol_process_memory_rss_total counter
otelcol_process_memory_rss_total 7.585792e+06
# HELP otelcol_process_runtime_heap_alloc_bytes_total Bytes of allocated heap objects (see 'go doc runtime.MemStats.HeapAlloc')
# TYPE otelcol_process_runtime_heap_alloc_bytes_total counter
otelcol_process_runtime_heap_alloc_bytes_total 5.40590824e+08
# HELP otelcol_process_runtime_total_alloc_bytes_total Cumulative bytes allocated for heap objects (see 'go doc runtime.MemStats.TotalAlloc')
# TYPE otelcol_process_runtime_total_alloc_bytes_total counter
otelcol_process_runtime_total_alloc_bytes_total 5.79787416e+08
# HELP otelcol_process_runtime_total_sys_memory_bytes_total Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')
# TYPE otelcol_process_runtime_total_sys_memory_bytes_total counter
otelcol_process_runtime_total_sys_memory_bytes_total 5.7855912e+08
# HELP otelcol_process_uptime_total Uptime of the process
# TYPE otelcol_process_uptime_total counter
otelcol_process_uptime_total 6.1469136692e+07
# HELP otelcol_target_info Target metadata
# TYPE otelcol_target_info gauge
otelcol_target_info{service_instance_id="2d918231-c9f3-414c-9aa6-f2c7fe382825",service_name="otelcorecol",service_version="0.65.0-dev"} 1
```

### CODE REVIEW

File: service/internal/proctelemetry/process_telemetry.go [hunk 1]
```diff
@@ -31,100 +48,224 @@ type processMetrics struct {
...
+	pm.otelSysMem, err = pm.meter.AsyncInt64().Counter(
+		"process_runtime_total_sys_memory_bytes",
+		instrument.WithDescription("Total bytes of memory obtained from the OS (see 'go doc runtime.MemStats.Sys')"),
+		instrument.WithUnit(unit.Bytes))
+	errors = multierr.Append(errors, err)
+
+	pm.otelCpuSeconds, err = pm.meter.AsyncFloat64().Counter(
+		"process_cpu_seconds",
+		instrument.WithDescription("Total CPU user and system time in seconds"),
+		instrument.WithUnit(unit.Milliseconds))
```
Comment (moh-osman3) [inline service/internal/proctelemetry/process_telemetry.go:157]: `instrument.WithUnit()` takes in a [unit.Unit](https://pkg.go.dev/go.opentelemetry.io/otel/metric@v0.34.0/unit#Unit) but the only provided unit for time is Millisecond.  The name of this metric is `process_cpu_seconds` so I'm wondering if I should rename this or if the metric name needs to be consistent with the OC metric name?

File: service/internal/proctelemetry/process_telemetry.go [hunk 2]
```diff
@@ -15,14 +15,31 @@
...
+	"go.opentelemetry.io/otel/metric/unit"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+
+	"go.opentelemetry.io/collector/featuregate"
+	"go.opentelemetry.io/collector/internal/obsreportconfig"
+)
+
+const (
+	scopeName      = "go.opentelemetry.io/collector/service/internal/proctelemetry"
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:41]: I'm not sure if this has to be 1 to 1 with the package path, but I think we may need a better name. Maybe: `go.opentelemetry.io/collector/service/process_telemetry` or something like that, I think needs to be a debate with the collector's folks.

File: service/internal/proctelemetry/process_telemetry.go (hunk 1 above)
Comment (paivagustavo) [reply service/internal/proctelemetry/process_telemetry.go:157]: I think for just leave out the unit, and add a comment about it. It is better to not have an unit than having the wrong one.

File: service/internal/proctelemetry/process_telemetry.go [hunk 3]
```diff
@@ -31,100 +48,224 @@ type processMetrics struct {
...
-		metric.WithUnit(stats.UnitSeconds))
-	if err != nil {
-		return err
-	}
-	if err = pm.processUptime.UpsertEntry(pm.updateProcessUptime); err != nil {
-		return err
-	}
+		ocmetric.WithDescription("Uptime of the process"),
+		ocmetric.WithUnit(stats.UnitSeconds))
+	errors = multierr.Append(errors, err)
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:213]: Although I like these changes, I would suggest you to not do them in this PR. Keep the changes to the current code as little as possible.

I would even suggest revert the change from `metric -> ocmetric` and instead rename the  otel `metric` to `otelmetric`

The less changes, the less conflict you'll have during reviews. Also think that this code will once be deleted after the migration!

File: service/internal/proctelemetry/process_telemetry.go [hunk 4]
```diff
@@ -31,100 +48,224 @@ type processMetrics struct {
...
-	}
+		ocmetric.WithDescription("Total physical memory (resident set size)"),
+		ocmetric.WithUnit(stats.UnitBytes))
+	errors = multierr.Append(errors, err)
 
-	return nil
+	return errors
+}
+
+func (pm *processMetrics) recordWithOC() error {
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:248]: As per the comment above, I would suggest us not changing these.

File: service/internal/proctelemetry/process_telemetry.go (hunk 1 above)
Comment (paivagustavo) [reply service/internal/proctelemetry/process_telemetry.go:157]: Actually, since Otel follows UCUM, [seconds should be represented as](https://unitsofmeasure.org/ucum#section-Base-Units): `s`. I think you should be able to do a `unit.Unit("s")`. 

We should ask otel-go to expand these as well.

File: service/internal/proctelemetry/process_telemetry.go [hunk 5]
```diff
@@ -31,100 +48,224 @@ type processMetrics struct {
...
+	processUptime *ocmetric.Float64DerivedCumulative
+	allocMem      *ocmetric.Int64DerivedGauge
+	totalAllocMem *ocmetric.Int64DerivedCumulative
+	sysMem        *ocmetric.Int64DerivedGauge
+	cpuSeconds    *ocmetric.Float64DerivedCumulative
+	rssMemory     *ocmetric.Int64DerivedGauge
+
+	// otel metrics
+	otelProcessUptime asyncfloat64.Counter
+	otelAllocMem      asyncint64.Counter
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:60]: we should keep the same metric types, can't change allocMem from gauge to counter as it yields wrong behavior and is not compatible.

File: service/internal/proctelemetry/process_telemetry.go [hunk 6]
```diff
@@ -31,100 +48,224 @@ type processMetrics struct {
...
+	otelTotalAllocMem asyncint64.Counter
+	otelSysMem        asyncint64.Counter
+	otelCpuSeconds    asyncfloat64.Counter
+	otelRssMemory     asyncint64.Counter
+
+	meter             otelmetric.Meter
+	logger            *zap.Logger
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:67]: why do we need this logger? It seems to be unused?

File: service/internal/proctelemetry/process_telemetry.go [hunk 7]
```diff
@@ -31,100 +48,224 @@ type processMetrics struct {
...
+	useOtelForMetrics bool
+	otelAttrs         []attribute.KeyValue
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:69]: same, seems unused?

**Review by paivagustavo — CHANGES_REQUESTED** [review]
Added a few comments but it is a great start! Thanks for taking this one :)

File: service/service.go [hunk 8]
```diff
@@ -211,9 +211,10 @@ func (srv *Service) initExtensionsAndPipeline(ctx context.Context, set Settings,
 		return fmt.Errorf("cannot build pipelines: %w", err)
 	}
 
-	if cfg.Telemetry.Metrics.Level != configtelemetry.LevelNone && cfg.Telemetry.Metrics.Address != "" {
-		// The process telemetry initialization requires the ballast size, which is available after the extensions are initialized.
-		if err = proctelemetry.RegisterProcessMetrics(srv.telemetryInitializer.ocRegistry, getBallastSize(srv.host)); err != nil {
+
+	if set.Config.Service.Telemetry.Metrics.Level != configtelemetry.LevelNone && set.Config.Service.Telemetry.Metrics.Address != "" {
```
Comment (paivagustavo) [inline service/service.go:215]: Also, what is the reason for changing how we get the telemetry level and address here?

File: service/service.go [hunk 9]
```diff
@@ -211,9 +211,10 @@ func (srv *Service) initExtensionsAndPipeline(ctx context.Context, set Settings,
...
+		processMetrics := proctelemetry.NewProcessMetrics(srv.telemetrySettings.Logger, srv.telemetryInitializer.registry, srv.telemetryInitializer.mp, getBallastSize(srv.host))
+		if err = processMetrics.RegisterProcessMetrics(context.Background(), srv.telemetryInitializer.ocRegistry); err != nil {
```
Comment (paivagustavo) [inline service/service.go:217]: I think it would be better if we could keep a similar function definition here and just pass down the feature gate registry + meterprovider, something like, no need to make this build a `newProcessMetrics`, just do it inside the `RegisterProcessMetrics`:
```
if err = proctelemetry.RegisterProcessMetrics(srv.telemetryInitializer.ocRegistry, getBallastSize(srv.host), meterProvider, registry); err != nil {
```

File: service/internal/proctelemetry/process_telemetry.go (hunk 2 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: Thanks for the review! Using your suggestion now.

File: service/internal/proctelemetry/process_telemetry.go (hunk 3 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: That makes sense to me, I will revert these changes.

File: service/internal/proctelemetry/process_telemetry.go (hunk 4 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: removed!

File: service/internal/proctelemetry/process_telemetry.go (hunk 5 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: Thanks for pointing this out changed the memory instruments to gauges!

File: service/internal/proctelemetry/process_telemetry.go (hunk 6 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: removed!

File: service/internal/proctelemetry/process_telemetry.go (hunk 7 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go:67]: removed!

File: service/service.go (hunk 8 above)
Comment (moh-osman3) [reply service/service.go]: Oh thanks for pointing this out. I think that was some rebase weirdness/mistake on my part. It actually breaks the code, so changed it back.

File: service/internal/proctelemetry/process_telemetry.go (hunk 1 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go]: Thanks for the suggestion! I am now using `unit.Unit("s")` Do you know how I can confirm that this worked? The output from prometheus doesn't seem to include the unit

e.g.
```
# HELP otelcol_process_uptime_total Uptime of the process
# TYPE otelcol_process_uptime_total counter
otelcol_process_uptime_total 10.295116
```

File: service/service.go (hunk 9 above)
Comment (moh-osman3) [reply service/service.go]: Got it I will take your suggestion in that case - it makes sense to minimize the amount of unnecessary changes i make in this pr.  I just really dislike having a function with too many arguments and it seemed like it was simple enough to avoid this. 

My thought process was to split the arguments into two groups. (1) arguments that will be fields of processMetric struct and can therefore be passed into `NewProcessMetrics()` and (2) arguments that are needed for metric recording that can be passed into `RegisterProcessMetrics()` i.e. the context and the ocRegistry. It also felt easier to test because creation of the processMetric struct could be done separately

File: service/internal/proctelemetry/process_telemetry.go [hunk 10]
```diff
@@ -127,6 +172,83 @@ func RegisterProcessMetrics(registry *metric.Registry, ballastSizeBytes uint64)
 	return nil
 }
 
+func (pm *processMetrics) createOtelMetrics() error {
+	var errors, err error
+
+	pm.otelProcessUptime, err = pm.meter.AsyncFloat64().Counter(
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:178]: I'm starting to think that for this one we can just have a simpler error just like the opencensus metrics
```
pm.otelProcessUptime, err = createOtelInstrument...
if err != nil {
	return err
}
err = registerCallback...
if err != nil {
	return err
}
```

we started this pattern of joining errors when the obsreport structs didn't return an error yet which is not the case here.

File: service/internal/proctelemetry/process_telemetry.go [hunk 11]
```diff
@@ -46,19 +72,38 @@ type processMetrics struct {
...
+
+		err = pm.recordWithOtel(ctx)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	pm.processUptime, err = ocRegistry.AddFloat64DerivedCumulative(
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:106]: I think it make sense to move the OC metrics to another function, so this RegisterProcessMetrics should be as simple as, this goes back to something similar to what you had earlier but in a simpler way :)
```go
	if pm.useOtelForMetrics {
		return pm.recordWithOtel(ctx)
	}

	return pm.recordWithOC(ctx)
```

File: service/internal/proctelemetry/process_telemetry.go (hunk 1 above)
Comment (paivagustavo) [reply service/internal/proctelemetry/process_telemetry.go]: Oh, you're right. The otel-go sdk doesn't export units yet, but I think that is cool for now.

File: service/service.go [hunk 12]
```diff
@@ -212,8 +212,7 @@ func (srv *Service) initExtensionsAndPipeline(ctx context.Context, set Settings,
 	}
 
 	if cfg.Telemetry.Metrics.Level != configtelemetry.LevelNone && cfg.Telemetry.Metrics.Address != "" {
-		// The process telemetry initialization requires the ballast size, which is available after the extensions are initialized.
```
Comment (paivagustavo) [inline service/service.go:215]: Was this comment removed intentionally?

**Review by paivagustavo — COMMENTED** [review]
Thanks for working through my comments, I've some small leftovers and it should be good to go!

File: service/internal/proctelemetry/process_telemetry.go (hunk 10 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go:172]: Thanks for pointing this out! I see now that for the obsreport PRs the `handleError` function was simply logging a warning and not returning the individual errors. Changed back to the original pattern.

File: service/internal/proctelemetry/process_telemetry.go (hunk 11 above)
Comment (moh-osman3) [reply service/internal/proctelemetry/process_telemetry.go:100]: I like this suggestion. Much simpler, thanks for the help!

File: service/service.go (hunk 12 above)
Comment (moh-osman3) [reply service/service.go:215]: That was an accident thanks for catching that!

File: service/internal/proctelemetry/process_telemetry.go [hunk 13]
```diff
@@ -127,6 +166,96 @@ func RegisterProcessMetrics(registry *metric.Registry, ballastSizeBytes uint64)
...
+		return err
+	}
+	err = pm.meter.RegisterCallback([]instrument.Asynchronous{pm.otelCpuSeconds}, func(ctx context.Context) {
+		pm.otelCpuSeconds.Observe(ctx, pm.updateCPUSeconds())
+	})
+	if err != nil {
+		return err
+	}
+
+	pm.otelRssMemory, err = pm.meter.AsyncInt64().Gauge(
```
Comment (paivagustavo) [inline service/internal/proctelemetry/process_telemetry.go:242]: nit: abbreviations should be all Uppercase if it is not in the begin of the variable name
```suggestion
	pm.otelRSSMemory, err = pm.meter.AsyncInt64().Gauge(
```

**Review by paivagustavo — APPROVED** [review]
LGTM, only one small comment  :)
